pymc3/dp/dp.py

```diff
@@ -0,0 +1,297 @@
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+import sys
+
+import warnings
+
+sys.path.insert(0, "../../../pymc3")
+
+import numpy as np
+import pandas as pd
+
+from aesara import tensor as at
+
+import pymc3 as pm
+
+import aesara
+from aesara.tensor.random.op import RandomVariable
+
+aesara.config.exception_verbosity = "high"
+
+from pymc3.math import logsumexp
+
+import argparse
+
+__all__ = [
+    "DirichletProcess",
+]
+
+parser = argparse.ArgumentParser()
+parser.add_argument("--dp", action="store_true")
+parser.add_argument("--dpmix", action="store_true")
+args = parser.parse_args()
+
+
+def stick_breaking(betas):
+    R"""
+    betas: vector of Beta random variables
+    """
+    # need to make sure that betas is one-dimensional
+
+    # if (betas.shape.eval()[0] <= 2):
+    #     raise AttributeError(
+    #         "More betas are needed to generate stick-breaking weights."
+    #     )
+
+    sticks = at.concatenate(
+        [
+            [1],
+            at.cumprod(1 - betas[:-1]),
+        ]
+    )
+
+    return betas * sticks
+
+
+def mixture_logp(weights, comp_dist, atoms, obs):
+    # comp_dist is something like lambda obs, atoms: pm.Normal.logp(obs, atoms, 1).eval()
+
+    atoms_shape = atoms.shape.eval()
+    obs_shape = obs.shape.eval()
+
+    # assert atoms.shape == weights.shape
+
+    weights = at.broadcast_to(weights, shape=obs_shape + atoms_shape)
+    log_weights = (at.log(weights)).T
+    atoms_logp = comp_dist(
+        at.broadcast_to(obs, atoms_shape + obs_shape),
+        at.broadcast_to(atoms, obs_shape + atoms_shape).T,
+    )
+
+    return logsumexp(log_weights + atoms_logp)
+
+
+class DirichletProcess:
+    R"""
+    Dirichlet Process Base class
+    """
+
+    def __init__(self, name, alpha, base_dist, K=None):
+        """
+        Examples
+        --------
+        .. code:: python
+
+            with pm.Model() as model:
+                alpha = pm.Gamma("concentration", 1, 1)
+                base_dist = pm.Normal("base_dist", 0, 3)
+                dp = DirichletProcess("dp", alpha, base_dist, 20)
+        """
+        if isinstance(alpha, (int, float)):
+            self.alpha = np.tile(alpha, reps=(K,))
+        elif isinstance(alpha, (np.ndarray, aesara.graph.basic.Variable)):
+            self.alpha = alpha
+        else:
+            raise ValueError(
+                "alpha parameter must be of type float, numpy.ndarray or "
+                f"pymc3.distributions.Distribution, but got {type(alpha)}"
+                "instead."
+            )
```

Comment:
I think we can simply do

```
betas = pm.Beta("betas", 1., self.alpha, size=K)
```

And broadcasting semantics are taken care automagically.